python/ray/experimental/channel/torch_tensor_type.py

```diff
@@ -71,110 +105,44 @@ def create_channel(
-            raise ValueError(
-                "DAG nodes wrapped with ray.experimental.TorchTensor must return a "
-                "torch.Tensor."
-            )
-        if tensor.shape != typ.shape:
-            raise ValueError(
-                "DAG node wrapped with ray.experimental.TorchTensor(shape="
-                f"{typ.shape}) returned "
-                f"a torch.Tensor of the shape {tensor.shape}"
-            )
-        if tensor.dtype != typ.dtype:
-            raise ValueError(
-                "DAG node wrapped with ray.experimental.TorchTensor(dtype="
-                f"{typ.dtype}) returned "
-                f"a torch.Tensor of the dtype {tensor.dtype}"
-            )
-
-        self.tensor = tensor
-        self.typ = typ
-
-
-class _TorchTensorSerializer:
-    def __init__(self, device: "torch.device"):
-        self.device = device
-
-    @staticmethod
-    def serialize_to_numpy(instance: "_TorchTensorWrapper") -> "np.ndarray":
-        tensor = instance.tensor
-        # Transfer through Ray's shared memory store for now.
-        # TODO(swang): This requires two copies, one to transfer from GPU to
-        # CPU and another from CPU to shared memory. Ideally we should elide
-        # the first copy and memcpy directly from GPU to the shared memory
-        # buffer.
-        if tensor.device.type == "cuda":
-            tensor = tensor.to("cpu")
-
-        return tensor.numpy()
 
-    def deserialize_from_numpy(self, np_array: "np.ndarray"):
+        # Transfer via host memory using a shared-memory channel.
         import torch
 
-        # TODO(swang): Support local P2P transfers if available.
-        # If there is a GPU assigned to this worker, move it there.
-        if self.device.type == "cuda":
-            # Use zero-copy from_numpy() because we are going to copy to GPU
-            # anyway.
-            # TODO: Pin the np_array memory to reduce data movement time.
-            # TODO: Set np_array.flags.writeable=True to avoid the PyTorch
-            # warning about not owning the underlying memory. This is safe to
-            # do as long as all other readers are also copying the data to a
-            # GPU.
-            cpu_tensor = torch.from_numpy(np_array)
-            return cpu_tensor.to(device=self.device)
-
-        # TODO(swang): Use zero-copy from_numpy() if np_array.flags.writeable
-        # is True. This is safe to set when deserializing np_array if the
-        # upstream task has num_readers=1.
-        return torch.tensor(np_array, device=self.device)
+        from ray.experimental.channel.shared_memory_channel import Channel
+
+        TORCH_DTYPE_ITEMSIZE_MAP = {
+            # INT types
+            torch.int: 4,
+            torch.uint8: 1,
+            torch.int8: 1,
+            torch.int32: 4,
+            torch.int64: 8,
+            torch.long: 8,
+            # FLOAT types
+            torch.half: 2,
+            torch.float: 4,
+            torch.float16: 2,
```

Comment:
Add bfloat16?

Reply:
Also probably should add other data types such as https://pytorch.org/docs/stable/tensors.html